controlpanel/api/rules.py

```diff
@@ -0,0 +1,194 @@
+add_perm('api.retrieve_app', is_authenticated & is_app_admin)
+add_perm('api.update_app', is_authenticated & is_superuser)
+add_perm('api.destroy_app', is_authenticated & is_superuser)
+add_perm('api.add_app_customer', is_authenticated & is_app_admin)
+add_perm('api.remove_app_customer', is_authenticated & is_app_admin)
+add_perm('api.add_app_admin', is_authenticated & is_superuser)
+add_perm('api.remove_app_admin', is_authenticated & is_superuser)
+add_perm('api.add_app_bucket', is_authenticated & is_superuser)
+add_perm('api.remove_app_bucket', is_authenticated & is_superuser)
+add_perm('api.view_app_logs', is_authenticated & is_app_admin)
+
+
+@predicate
+def has_bucket_access(user, obj, **kwargs):
+    """
+    Check whether `user` has access to;
+    - the `obj` `S3Bucket`
+    - the `s3bucket` of the `obj` `AppS3Bucket`
+    - the `s3bucket` of the `obj` `UserS3Bucket`
+
+    :param user User: The user to check
+    :param obj AppS3Bucket|S3Bucket|UserS3Bucket: The object to check
+    :param kwargs dict: Additional search criteria
+    """
+    if obj is None:
+        return True
+
+    if isinstance(obj, S3Bucket):
+        try:
+            UserS3Bucket.objects.get(user=user, s3bucket=obj, **kwargs)
+        except UserS3Bucket.DoesNotExist:
+            return False
+        return True
+
+    if isinstance(obj, AppS3Bucket) or isinstance(obj, UserS3Bucket):
+        return has_bucket_access(user, obj.s3bucket)
+
+    # XXX raise exception?
+    return False
+
+
+@predicate
+def has_bucket_write_access(user, obj):
+    return has_bucket_access(user, obj, access_level=UserS3Bucket.READWRITE)
+
+
+@predicate
+def is_bucket_admin(user, obj):
+
+    if obj is None:
+        return True
+
+    if isinstance(obj, S3Bucket):
+        return has_bucket_access(user, obj, is_admin=True)
+
+    if isinstance(obj, AppS3Bucket) or isinstance(obj, UserS3Bucket):
+        return is_bucket_admin(user, obj.s3bucket)
+
+    # XXX raise an exception?
+    return False
+
+
+add_perm('api.list_s3bucket', is_authenticated)
+add_perm('api.create_s3bucket', is_authenticated)
+add_perm('api.retrieve_s3bucket', is_authenticated & has_bucket_access)
+add_perm('api.update_s3bucket', is_authenticated & has_bucket_write_access)
+add_perm('api.destroy_s3bucket', is_authenticated & is_bucket_admin)
+add_perm('api.add_s3bucket_admin', is_authenticated & is_superuser)
+add_perm('api.remove_s3bucket_admin', is_authenticated & is_superuser)
+add_perm('api.grant_s3bucket_access', is_authenticated & is_bucket_admin)
+
+
+@predicate
+def is_self(user, other):
+    """
+    Check whether `user` is performing an action on themselves
+
+    :param user User: The user to check
+    :param other User: The other user to check
+    """
+    if other is None:
+        return True
+
+    return user == other
+
+
+add_perm('api.list_user', is_authenticated)
+add_perm('api.create_user', is_authenticated & is_superuser)
+add_perm('api.retrieve_user', is_authenticated & is_self)
```

Comment:
How does `is_self` works here?

Is `add_perm()` passing `other` somehow or `other` is always `None` and `is_self()` always return `True`?

Reply:
When we call `user.has_perm('api.retrieve_user', other_user)`, the auth backend first checks `is_self(request.user)` and if the result would be `False` (which it never is, because it doesn't make sense), we would get `PermissionDenied`. Otherwise, it then calls `is_self(request.user, other_user)`.